src/components/tl/cuda/tl_cuda_team_topo.c

```diff
@@ -2,6 +2,200 @@
 #include "tl_cuda.h"
 
 #define UCC_TL_CUDA_TEAM_TOPO_SAME_DEVICE -1
+#define UCC_TL_CUDA_TEAM_TOPO_MAX_RINGS 32
+
+static ucc_status_t
+ucc_tl_cuda_team_topo_add_ring(const ucc_tl_cuda_team_t *team,
+                               ucc_tl_cuda_team_topo_t *topo,
+                               ucc_tl_cuda_ring_t *ring,
+                               int invert, int num_dups)
+{
+    ucc_rank_t size = UCC_TL_TEAM_SIZE(team);
+    ucc_tl_cuda_ring_t *new_ring;
+    ucc_status_t status;
+    int i, j;
+
+    for (i = 0; i < num_dups; i++) {
+        topo->rings[topo->num_rings + i].ring  = NULL;
+        topo->rings[topo->num_rings + i].iring = NULL;
+    }
+
+    for (i = 0; i < num_dups; i++) {
+        new_ring = &topo->rings[topo->num_rings + i];
+        new_ring->ring  = (ucc_rank_t*)ucc_malloc(size * sizeof(ucc_rank_t));
```

Comment:
made single allocation for ring and iring, but sharing same allocation for multiple rings is tricky since in topo_destory need to know what ring was allocated first and what is duplicated